### ISSUE TITLE
Calling an asynchronous method in the ReceiveAsync method may cause deadlock

### ISSUE THREAD

**Issue by ClockGet** [akkadotnet/akka.net#5067]
Akka.Net version: 1.4.20
platform: win 10, .net core 3.1

code:
```
public class TestAsync
{
	public static void Test()
	{
		using var system = ActorSystem.Create("testAsync");
		var sender = system.ActorOf(Props.Create<TestActor>(), "Test");
		sender.Tell(new TestAsyncMessage());
		Console.ReadKey();
	}
}
internal class TestAsyncMessage { }
internal class TestActor : ReceiveActor
{
	public TestActor()
	{
		ReceiveAsync<TestAsyncMessage>(async t =>
		{
			await Task.Delay(1000);
			Test();
		});
	}
	void Test()
	{
		AsyncFunc().Wait();
	}
	async Task AsyncFunc()
	{
		Console.WriteLine("before async");
		await Task.Delay(1000);
		Console.WriteLine("after async");
	}
}
```
**Actual behavior**
The console only prints "before async" and then nothing happens.
**Expected behavior**
The console can prints "before async" and "after async".

After investigation, I found the current **TaskScheduler** when the AsyncFunc method is called is **Akka.Dispatch.ActorTaskScheduler**,
and we know that the key point behind the await keyword is the **TaskAwaiter** class, and the Task instance method **SetContinuationForAwait** is called in the **OnCompletedInternal** method of the **TaskAwaiter** class before calling the rest of the code(Console.WriteLine("after async");). In this method, it tries to get TaskScheduler.InternalCurrent and make the rest of the code(Console.WriteLine("after async");) be scheduled by "InternalCurrent". 
In addition, when I added ConfigureAwait(false) to the end of await, the deadlock disappeared, because ConfigureAwait(false) would ignore the current TaskScheduler.
I know that in the era of .net framework, calling the Wait method on the async method is an important cause of deadlock, but now it is the era of dotnet core. Please give me some suggestions, thanks!

**Comment by to11mtm** [comment]
> I know that in the era of .net framework, calling the Wait method on the async method is an important cause of deadlock, but now it is the era of dotnet core. Please give me some suggestions, thanks!

This is a fairly common misconception about .NET Core, but [the truth is more nuanced](https://devblogs.microsoft.com/dotnet/configureawait-faq/#ive-heard-configureawaitfalse-is-no-longer-necessary-in-net-core-true). ASPNETCORE on it's own no longer has it's own `SynchronizationContext`, However other applications/libraries certainly can and do still depend on `SynchronizationContext`, and this includes Akka.Net

**Comment by ClockGet** [comment]
@to11mtm But in `ReceiveAsync` scope, `SynchronizationContext.Current` return `null`, and this problem is not caused by `SynchronizationContext` either.

**Comment by to11mtm** [comment]
@ClockGet That is true, I should have been a little more detailed in my answer.

In addition to `SynchronizationContext`, Custom `TaskScheduler` implementations can still result in the requirement of `.ConfigureAwait(false)`. This does include Akka.NET (which does use custom `TaskScheduler`s.), and I believe it still can be a hazard in WPF and WinForms.

**Comment by Aaronontheweb** [comment]
Believe this has been fixed with some of the work we did in v1.5, but it'll take some more effort in future minor version releases to fully resolve this.
